File: code-0.10.27/apps/app/src/api/metadata.rs
use crate::api::Result;
use daedalus::minecraft::VersionManifest;
use daedalus::modded::Manifest;
use theseus::metadata::{
    BedrockContentDetails, BedrockContentPage, BedrockInstalledContentEntry,
    BedrockVersionEntry,
};

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("metadata")
        .invoke_handler(tauri::generate_handler![
            metadata_get_game_versions,
            metadata_get_loader_versions,
            metadata_get_bedrock_versions,
            metadata_get_bedrock_content,
            metadata_get_bedrock_content_details,
            metadata_install_bedrock_content,
            metadata_get_installed_bedrock_content,
        ])
        .build()
}

/// Gets the game versions from daedalus
#[tauri::command]
pub async fn metadata_get_game_versions() -> Result<VersionManifest> {
    Ok(theseus::metadata::get_minecraft_versions().await?)
}

/// Gets the fabric versions from daedalus
#[tauri::command]
pub async fn metadata_get_loader_versions(loader: &str) -> Result<Manifest> {
    Ok(theseus::metadata::get_loader_versions(loader).await?)
}

/// Gets Bedrock versions from the remote LiteLDev version database
#[tauri::command]
pub async fn metadata_get_bedrock_versions() -> Result<Vec<BedrockVersionEntry>>
{
    Ok(theseus::metadata::get_bedrock_versions().await?)
}

#[tauri::command]
pub async fn metadata_get_bedrock_content(
    kind: &str,
    page: Option<u32>,
) -> Result<BedrockContentPage> {
    Ok(theseus::metadata::get_bedrock_content(kind, page).await?)
}

#[tauri::command]
pub async fn metadata_get_bedrock_content_details(
    kind: &str,
    page_url: &str,
) -> Result<BedrockContentDetails> {
    Ok(
        theseus::metadata::get_bedrock_content_details(kind, page_url)
            .await?,
    )
}

#[tauri::command]
pub async fn metadata_install_bedrock_content(
    profile_path: &str,
    kind: &str,
    page_url: &str,
) -> Result<String> {
    Ok(
        theseus::metadata::install_bedrock_content(
            profile_path,
            kind,
            page_url,
        )
        .await?,
    )
}

#[tauri::command]
pub async fn metadata_get_installed_bedrock_content(
    profile_path: &str,
) -> Result<Vec<BedrockInstalledContentEntry>> {
    Ok(
        theseus::metadata::get_installed_bedrock_content(profile_path)
            .await?,
    )
}

File: apps/app-frontend/src/helpers/metadata.js
import { invoke } from '@tauri-apps/api/core'

/// Gets the game versions from daedalus
// Returns a VersionManifest
export async function get_game_versions() {
	return await invoke('plugin:metadata|metadata_get_game_versions')
}

// Gets the given loader versions from daedalus
// Returns Manifest
export async function get_loader_versions(loader) {
	return await invoke('plugin:metadata|metadata_get_loader_versions', { loader })
}

// Gets Bedrock versions as tuples: [version, packageId, 0]
export async function get_bedrock_versions() {
	return await invoke('plugin:metadata|metadata_get_bedrock_versions')
}

// Gets Bedrock content list from mcpehub
// kind: 'textures' | 'shaders' | 'maps'
export async function get_bedrock_content(kind, page = 1) {
	return await invoke('plugin:metadata|metadata_get_bedrock_content', {
		kind,
		page,
	})
}

export async function get_bedrock_content_details(kind, pageUrl) {
	return await invoke('plugin:metadata|metadata_get_bedrock_content_details', {
		kind,
		pageUrl,
	})
}

// Downloads and installs a Bedrock content item into profile files
export async function install_bedrock_content(profilePath, kind, pageUrl) {
	return await invoke('plugin:metadata|metadata_install_bedrock_content', {
		profilePath,
		kind,
		pageUrl,
	})
}

export async function get_installed_bedrock_content(profilePath) {
	return await invoke('plugin:metadata|metadata_get_installed_bedrock_content', {
		profilePath,
	})
}
